kymatio/scattering1d/filter_bank.py

```diff
@@ -712,22 +823,675 @@ def scattering_filter_factory(J_support, J_scattering, Q, r_psi=math.sqrt(0.5),
+            # J_pad_frs_max_init
+            compute_all_subsamplings(phi_f_fr, j0=j0)
+        else:
+            # These won't differ from plain subsampling but we still index
+            # via `subsample_equiv_relative_to_max_pad_init` and
+            # `n1_fr_subsample` so just copy pointers.
+            # `phi[::factor] == gauss_1d(N // factor, sigma_low * factor)`
+            # when not aliased
+            phi_f_fr[j0] = [phi_f_fr[0][j0_sub]
+                            for j0_sub in range(j0, 1 + log2_F)]
+
+    # embed meta info in filters
+    phi_f_fr.update({field: {} for field in ('xi', 'sigma', 'j', 'width',
+                                             'support')})
+    j0s = [j for j in phi_f_fr if isinstance(j, int)]
+    for j0 in j0s:
+        xi_fr_0 = 0.
+        sigma_fr_0 = (sigma_low if sampling_phi_fr == 'resample' else
+                      sigma_low * 2**j0)
+        j0_0 = (log2_F if sampling_phi_fr == 'resample' else
+                log2_F - j0)
+        for field in ('xi', 'sigma', 'j', 'support', 'width'):
+            phi_f_fr[field][j0] = []
+        phi_f_fr['xi'][j0] = xi_fr_0
+        phi_f_fr['sigma'][j0] = sigma_fr_0
+        phi_f_fr['j'][j0] = j0_0
+        phi_f_fr['width'][j0] = []
+        phi_f_fr['support'][j0] = []
+        for j0_sub in range(len(phi_f_fr[j0])):
+            # should halve with subsequent j0_sub, but compute exactly.
+            # `j0`-to-`N_fr` uniqueness asserted in `psi_fr_factory`
+            phi = phi_f_fr[j0][j0_sub]
+            width = compute_temporal_width(
+                phi, N=len(phi), sigma0=sigma0,
+                criterion_amplitude=criterion_amplitude)
+            support = 2*compute_temporal_support(
+                phi, criterion_amplitude=criterion_amplitude)
+            phi_f_fr['width'][j0].append(width)
+            phi_f_fr['support'][j0].append(support)
 
     # return results
-    return phi_f, psi1_f, psi2_f, t_max_phi
+    return phi_f_fr
+
+#### Energy renormalization ##################################################
+def energy_norm_filterbank_tm(psi1_f, psi2_f, phi_f, J, log2_T):
+    """Energy-renormalize temporal filterbank; used by `base_frontend`.
+    See `help(kymatio.scattering1d.filter_bank.energy_norm_filterbank)`.
+    """
+    pass
+
+
+def energy_norm_filterbank_fr(psi1_f_fr_up, psi1_f_fr_down, phi_f_fr,
+                              J_fr, log2_F):
+    """Energy-renormalize frequential filterbank; used by `base_frontend`.
+    See `help(kymatio.scattering1d.filter_bank.energy_norm_filterbank)`.
+    """
+    pass
+
+
+#### misc / long #############################################################
+def conj_fr(x):
+    """Conjugate in frequency domain by swapping all bins (except dc);
+    assumes frequency along first axis.
+    """
+    out = np.zeros_like(x)
+    out[0] = x[0]
+    out[1:] = x[:0:-1]
+    return out
+
+def compute_temporal_width(p_f, N=None, pts_per_scale=6, fast=True,
```

Comment:
> The last four keyword arguments are unused!

No, only `pts_per_scale` is unused, and that I left since it helps with prototyping if necessary - but I'm fine with hiding it.

> Do we really need two functions

Yes. Refer to our Discord discussions, also [here](https://github.com/kymatio/kymatio/discussions/730#discussioncomment-1018400)

Reply:
It's still not clear to me what the chain of dependencies is here, between user-defined T and filter design. It's also not clear what sigma0 does in this context.

At the end of the day, what the user wants is to design a representation which guarantees a certain amount of invariance T. Your function does the opposite: measure the invariance of a filter that's already designed.

If the invariance is not feasible because the filter can't be designed (signal too short), we shouldn't hesitate to raise a helpful error message.

Reply:
(although, to your credit, this is not possible in frequency scattering, in which case we should indeed pad the frequency domain to allow a filter of width F being constructible)

Reply:
User wants `T`. How do we know we're giving it to them? Continuous-time expressions guarantee nothing: we must measure to validate. The `width` meta is used directly in JTFS with `sampling_psi_fr='exclude'`.

`sigma0` is how `T` maps to final `sigma` for `phi`. It's always there, even if it's not there (in which case it's just `1`: `sigma=1/T`).

Reply:
From what you're saying, i understand that `sigma0` should be computed at runtime from the specification of `T` and `phi`. Changing `sigma0` defeats the purpose of adjusting `phi` according to an objective definition of invariance `T=phi.width`. It should not be a keyword argument anymore.

> The width meta is used directly in JTFS with sampling_psi_fr='exclude'

what's the criterion for `exclude`? i imagine something involving `psi2` and `psi_fr`?

Reply:
```python
if scf.sampling_psi_fr == 'exclude':
    width = psi1_f[n1_fr]['width'][subsample_equiv_due_to_pad]
    if width > scf.N_frs[n2]:
    	continue
```
If `psi_fr` is too wide relative to input -> skip. You have something similar in `scattering.m`